# 문자열(String)



---



## 문자열 정의

* 문자, 단어 등으로 구성된 문자들의 집합
* 시퀀스 자료형(sequence data type)



In [2]:
s = "String"
print(s)
print(s[0])  # 인덱스는 0부터 시작
print(s[1])
print(s[2])
print(s[3])
print(s[4])
print(s[5])

String
S
t
r
i
n
g


In [3]:
s = "문자열"
print(s)
print(s[0])
print(s[1])
print(s[2]) # 유니코드 지원하므로 한국어도 문자열 처리 가능!

문자열
문
자
열


## 문자열 생성

* 파이썬에서 문자열은 작은따옴표(') 또는 큰 따옴표(")로 표현


In [4]:
print('Hello')  # 싱글 쿼테이션
print("파이썬은 재미있다.")  # 더블 쿼테이션
print('''파이썬은 심플하다.''')  # 싱글 쿼테이션 3개
print("""파이썬은 문자열 처리가 뛰어나다.""")  # 더블 쿼테이션 3개

Hello
파이썬은 재미있다.
파이썬은 심플하다.
파이썬은 문자열 처리가 뛰어나다.


### 따옴표가 있는 문자열 생성

* 문자열에 작은따옴표가 있을 경우, 큰따옴표로 둘러싸서 표현
* 문자열에 큰따옴표가 있을 경우, 작은따옴표로 둘러싸서 표현
* 이스케이프 코드 \를 이용하여 작은따옴표(‘)와 큰따옴표(“)를 문자열에 포함

In [8]:
# 작은따옴표 있을 경우 더블쿼테이션으로 묶어줌
string = "Python's built-in string classes" 
print(string)

# 큰따옴표 있을 경우 싱글쿼테이션으로 묶어줌
string = '모두가 "파이썬을 매우 쉽다."라고 말한다.'
print(string)

# 백슬래쉬(이스케이프)문자로 따옴표 인식
string = 'Python\'s built-in string classes' 
print(string)

# 백슬래쉬(이스케이프)문자로 따옴표 인식
string = "모두가 \"파이썬을 매우 쉽다.\"라고 말한다."
print(string)

Python's built-in string classes
모두가 "파이썬을 매우 쉽다."라고 말한다.
Python's built-in string classes
모두가 "파이썬을 매우 쉽다."라고 말한다.


### 이스케이프 문자

* 특수 문자
* 문자 앞에 백슬래쉬(\) 사용
* 문자열에서 특수 문자를 표현하는데 사용


### 여러 줄이 있는 문자열 생성

* 이스케이프 문자(\\n)를 이용하여 여러 줄이 있는 문자열 생성 가능

In [9]:
text = "동해물과 백두산이 마르고 닳도록\n하느님이 보우하사 우리나라 만세\n무궁화 삼천리 화려강산\n대한사람 대한으로 길이 보전하세"
print(text)

동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산
대한사람 대한으로 길이 보전하세


* 작은 따옴표 3개 또는 큰 따옴표 3개를 이용하여 여러 줄이 있는 문자열 생성

In [10]:
text = """동해물과 배두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세.
무궁화 삼천리 화려강산
대한사람 대한으로 길이 보전하세"""
print(text)

동해물과 배두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세.
무궁화 삼천리 화려강산
대한사람 대한으로 길이 보전하세




---



## 문자열 연산(String Operators)

### 문자열 더하기

+ 연산자를 사용하여 문자열 연결


In [11]:
s1 = "동해물과 "
s2 = "백두산이"

print(s1 + s2)

동해물과 백두산이


### 문자열 곱하기

* \* 연산자를 사용하여 문자열 반복


In [12]:
s = 'String'
print(s * 3)

StringStringString


### 문자열 길이(length)

* 문자열 길이를 구하는 `len()` 함수

In [15]:
s = 'String '
print(len(s))
print(len(s * 3))

7
21


### 문자열 인덱싱(indexing)

* 문자열은 리스트처럼 문자 하나하나가 상대적인 주소(offset)를 가짐
* 이 주소를 사용해 할당된 값을 가져오는 인덱싱을 사용

In [16]:
s = "String"
print(s)
print(s[-6])
print(s[-5])
print(s[-4])
print(s[-3])
print(s[-2])
print(s[-1])

String
S
t
r
i
n
g


### 문자열 슬라이싱(slicing)

* 문자열의 주소를 이용하여 문자열을 조각(부분)을 추출

In [20]:
s = "HyoungsunLab-computer lab"
print(s)
print(s[0:7])
print(s[:7])
print(s[10:])
print(s[-10:])
print(s[-19:-11])

HyoungsunLab-computer lab
Hyoungs
Hyoungs
ab-computer lab
mputer lab
sunLab-c




---



## 문자열 메소드(String Methods)

* 파이썬은 문자열 처리를 아주 쉽게 처리할 수 있는 많은 메소드들을 제공
* 문자열을 다루는 주요 메소드들을 표로 정리


### capitalize() / casefold()

* 문자열 "string"에 대해 `capitalize()`를 실행하면, 첫 문자가 대문자인 "String" 문자열로 변환
* 문자열 "String"에 대해 `casefold()`를 실행하면, 모든 문자가 소문자인 "string" 문자열로 변환

In [23]:
s = "string"
print(s)
s = s.capitalize()
print(s)
s = s.casefold()
print(s)

string
String
string


### count()

* 문자열 "string string"에 포함된 문자 's'의 갯수를 반환하도록 `count()`를 실행
* 문자 's'가 총 2개 존재하므로 2를 반환
* 부분 문자열 "str"이 몇개인지 반환하는 것도 가능하고, 부분 문자열 "str"이 총 2개 존재하므로 2를 반환

In [25]:
s = "string string"
print(s)
print(s.count('s'))    # 문자의 개수
print(s.count('str'))  # 문자열의 개수

string string
2
2


### find() / rfind()

* 문자열 "string string"에서 문자 's'의 해당 위치를 `find()`를 통해 반환
* 문자열 "string string"에서 부분 문자열 "ing"의 해당 위치를 `find()`를 통해 반환
* 문자열 "string string"에서 문자 's'를 해당 위치를 `rfind()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환
* 문자열 "string string"에서 부분 문자열 "ring"를 해당 위치를 `rfind()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환

In [30]:
s = 'string string'
print(s.find('s'))  # 's'의 위치 인덱스값 반환
print(s.find('ing')) # 왼쪽부터 찾음
print(s.rfind('s'))
print(s.rfind('ring'))  # 오른쪽부터 찾기

0
3
7
9


### index() / rindex()

* 문자열 "string string"에서 문자 's'의 해당 위치를 `index()`를 통해 반환
* 문자열 "string string"에서 부분 문자열 "ing"의 해당 위치를 `index()`를 통해 반환
* 문자열 "string string"에서 문자 's'를 해당 위치를 `rindex()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환
* 문자열 "string string"에서 부분 문자열 "ring"를 해당 위치를 `rindex()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환
* 문자열 "string string"에서 문자 'z'를 해당 위치를 `rindex()` 함수를 통해 오른쪽부터 탐색하지만 찾을 수 없어서 'ValueError' 발생

In [32]:
s = 'string string'
print(s.index('s'))  # 's'의 위치 인덱스값 반환
print(s.index('ing')) # 왼쪽부터 찾음
print(s.rindex('s'))
print(s.rindex('ring'))  # 오른쪽부터 찾기
# 인덱스 없으면 ValueError 발생시킴

0
3
7
9


### isalnum()

* `isalnum()`은 문자열에 알파벳이나 숫자가 1개 이상 있으면 True 반환
* 문자열 "string"은 알파벳으로 구성 되었으므로 True 반환
* 문자열 "한글"은 알파벳으로 구성 되어있으므로 True 반환
* 문자열 "!@#"은 특수기호로 구성 되어있으므로 False 반환
* 문자열 "123"은 숫자들로 구성 되어있으므로 True 반환

In [34]:
print('string'.isalnum())
print('한글'.isalnum())
print('!@#'.isalnum())  # 특수문자는 False
print('123'.isalnum())

True
True
False
True


### isalpha()

* `isalpha()`는 문자열에 알파벳이 1개 이상 있으면 True 반환
* 문자열 "string"은 알파벳으로 구성 되었으므로 True 반환
* 문자열 "한글"은 알파벳으로 구성 되어있으므로 True 반환
* 문자열 "!@#"은 특수기호로 구성 되어있으므로 False 반환
* 문자열 "123"은 숫자들로 구성 되어있으므로 Flase 반환

In [35]:
print('string'.isalpha())
print('한글'.isalpha())
print('!@#'.isalpha())  # 특수문자는 False
print('123'.isalpha())  # 숫자도 False

True
True
False
False


### isdecimal()

* `isdecimal()`는 문자열의 모든 문자가 10진수 문자이면 True 반환
* 문자열 "123"은 모두 10진수 문자열이므로 True 반환
* 문자열 "1.23"은 실수형 문자열이므로 False 반환

In [37]:
print('123'.isdecimal())
print('1.23'.isdecimal()) # 실수이므로 False

True
False


### isdigit()

* `isdigit()`는 문자열의 모든 문자가 숫자일 때 True 반환
* 문자열 "123"은 모든 문자가 숫자에 해당하므로 True 반환
* 문자열 "1.23"은 문자 중 숫자에 해당안되는 . 이 존재하여 False

In [38]:
print('123'.isdigit())
print('1.23'.isdigit()) # .이 있어서 False

True
False


### isnumeric()

* `isnumeric()`는 문자열의 모든 문자가 수치형일 때 True 반환
* 문자열 "123"은 모든 문자가 숫자에 해당하므로 True 반환
* 문자열 "1.23"은 문자 중 수치형에 해당안되는 . 이 존재하여 False

In [39]:
print('123'.isnumeric())
print('1.23'.isnumeric()) # 여기도 점 있으므로 False

True
False


### isidentifier()

* `isidentifier()`는 문자열이 파이썬에서 사용하는 식별자인 경우 True 반환
* 문자열 "123"은 식별자로 사용되지 않는 문자열이라서 False 반환
* 문자열 "True"는 식별자로 사용되기 때문에 True 반환
* 문자열 "print"는 식별자로 사용되기 때문에 True 반환

In [41]:
print('123'.isidentifier())
print('True'.isidentifier())
print('print'.isidentifier())

False
True
True


### isspace()

* 문자열 " "은 문자열 내에 공백 문자가 있어서 True 반환
* 문자열 " 1 "은 문자열 내에 공백 아닌 문자가 있어서 False 반환

In [42]:
print(' '.isspace())
print(' 1 '.isspace()) # 공백 아닌 숫자 포함되면 False

True
False


### istitle()

* 문자열 "String"은 문자열 내에 제목과 같이 첫 글자가 대문자여서 True 반환
* 문자열 "STRING"은 모든 문자가 대문자이기 때문에 False 반환

In [43]:
print('String'.istitle())  # 첫 글자가 대문자여야 함
print('STRING'.istitle())

True
False


### islower()

* 문자열 "string"은 모든 문자가 소문자여서 True 반환
* 문자열 "String"은 첫 문자가 대문자여서 False 반환

In [44]:
print('string'.islower())
print('String'.islower())

True
False


### isupper() 함수

* 문자열 "STRING"은 모든 문자가 대문자여서 True 반환
* 문자열 "String"은 첫 문자만 대문자여서 False 반환

In [45]:
print('STRING'.isupper())
print('String'.isupper())

True
False


### join()

* 문자열 "String"에 대해서 `join()`으로 공백 문자 ' '를 문자 사이마다 추가한 문자열 "S t r i n g"을 반환
* 문자열 "String"에 대해서 `join()` 으로 공백 문자 '_'를 문자 사이마다 추가한 문자열 "S_t_r_i_n_g"을 반환
* 문자열 "String"에 대해서 `join()` 으로 공백 문자 '|'를 문자 사이마다 추가한 문자열 "S|t|r|i|n|g"을 반환

In [47]:
s = 'String'
print(' '.join(s)) # 공백이 각 문자 사이사이마다 들어감(join)
print('_'.join(s))
print('|'.join(s))

S t r i n g
S_t_r_i_n_g
S|t|r|i|n|g


### center() / ljust() / rjust()

* 문자열 "String"에 대해서 `center()`를 사용하여 너비 10에 해당하는 문자열 길이에 가운데 정렬한 문자열 '  String  '로 변환
* 문자열 "String"에 대해서 `ljust()`를 사용하여 너비 10에 해당하는 문자열 길이에 왼쪽 정렬한 문자열 'String    '로 변환
* 문자열 "String"에 대해서 `rjust()`를 사용하여 너비 10에 해당하는 문자열 길이에 오른쪽 정렬한 문자열 '    String'로 변환

In [49]:
print("'" + "String".center(10) + "'")
print("'" + "String".ljust(10) + "'")
print("'" + "String".rjust(10) + "'")

'  String  '
'String    '
'    String'


### lower() / upper() / title() / swapcase()

* 문자열 "String"을 `lower()`를 이용하여 모든 문자가 소문자로 변환된 "string" 반환
* 문자열 "String"을 `upper()`를 이용하여 모든 문자가 대문자로 변환된 "STRING" 반환
* 문자열 "string"을 `title()`를 이용하여 첫 문자가 대문자로 변환된 "String" 반환
* 문자열 "String"을 `swapcase()`를 이용하여 대문자는 소문자로 변환되고 소문자는 대문자로 변환된 "sTRING" 반환


In [50]:
print('String'.lower())
print('String'.upper())
print('string'.title())
print('String'.swapcase())

string
STRING
String
sTRING


### strip() / lstrip() / rstrip()

* 양쪽 공백이 포함된 문자열 "  String  "을 `strip()`를 통해 공백 제거
* 양쪽 공백이 포함된 문자열 "  String  "을 `lstrip()`를 통해 왼쪽 공백 제거
* 양쪽 공백이 포함된 문자열 "  String  "을 `rstrip()`를 통해 오른쪽 공백 제거

In [51]:
print("     String   ".strip())
print("     String   ".lstrip())  # 왼쪽 공백 제거
print("     String   ".rstrip())  # 오른쪽 공백 제거

String
String   
     String


### partition() / rpartition()

* 문자열 "String"을 `partition()`를 이용하여 문자 't' 기준으로 분할
* 문자열 "String String"을 `rpartition()`를 이용하여 마지막 문자 'S'를 기준으로 분할

In [53]:
print('String'.partition('t'))
print('String String'.rpartition('s'))  # 오른쪽에서 탐색

('S', 't', 'ring')
('', '', 'String String')


### replace()

* 문자열 "String"을 `replace()`를 이용해 문자열 "Str"을 문자 'R'로 교체

In [55]:
print('String'.replace('Str', 'R'))

Ring


### split() / rsplit() / splitlines()

* 문자열 "1 2 3"을 `split()`의 기본 구분자를 이용하여 '1', '2', '3' 으로 구분
* 문자열 "1_2_3"을 `split()`으로 구분자 '_'를 이용하여 '1', '2', '3' 으로 구분
* 문자열 "1 2 3"을 `rsplit()`으로 구분자 '_'를 이용하여 오른쪽부터 1번만 구분하여 '1_2', '3' 으로 구분
* 문자열 "123\n123\n123\n"을 `splitlines()` 함수를 이용하여 라인 단위로 구분하여 '123, '123, '123 반환

In [60]:
print("1 2 3".split()) # 기본 구분자 : 공백
print("1_2_3".split('_')) # 구분자 지정해줌
print("1_2_3".rsplit('_', 1))  # 오른쪽부터 1번만 구분해달라는 뜻
print("123\n123\n123\n".splitlines()) # 줄단위로 구분

['1', '2', '3']
['1', '2', '3']
['1_2', '3']
['123', '123', '123']


### startswith() / endswith()

* 문자열 "String"에서 시작 문자가 'S'인지 `startswith()`를 통해 확인
* 문자열 "String"에서 마지막 문자가 'g'인지 `endswith()`를 통해 확인

In [62]:
print('String'.startswith('S')) # 대문자 S로 시작해서 True
print('String'.endswith('g'))

True
True


### zfill()

* 문자열 "123"에서 너비 8만큼으로 늘리고 비어있는 부분에 '0'이 채워진 문자열 "00000123" 반환

In [64]:
print("123".zfill(8)) # 너비 8만큼 늘리고 0으로 채워줘.

00000123




---



## 문자열 서식(String Format)


### 문자열 포맷팅(String Formatting)

* 문자열 내에서 서식에 맞추어 특정 값을 삽입 또는 변경
* 기호 % 뒤에 있는 값이 문자열 내의 서식에 순서대로 매핑

In [68]:
print("나는 사과가 %d개 있다" % 2)
print("나는 사과가 %s개 있다" % '두')
print("나는 %s가 %d개 있다" % ('사과', 2))

나는 사과가 2개 있다
나는 사과가 두개 있다
나는 사과가 2개 있다


### 정렬, 공백, 소수점 포맷

* 포맷 문자 앞에 숫자는 길이를 의미
* -는 왼쪽 정렬을 의미
* 소수점 ‘.’ 뒤에 숫자는 소숫점 이하 개수를 의미

In [73]:
print('%8s' % "Hello") # 오른쪽 정렬, 8칸
print("%-8sPython" % "Hello") # 왼쪽 정렬
print("%0.2f" % 3.14159265)  # 소수점 2자리까지만 출력
print("%8.2f" % 3.14159265)  # 소수점 2자리까지만 출력
print("%-8.2f" % 3.14159265)  # 소수점 2자리까지만 출력

   Hello
Hello   Python
3.14
    3.14
3.14    


### 문자열 format 메소드

* 문자열 `format()`의 인덱스를 이용한 고급 포맷팅
* 인덱스 번호에 따라 `format()` 함수의 각 값들이 매핑

In [75]:
print("나는 사과가 {0}개 있다.".format(2))
print("나는 사과가 {0}개 있다.".format('두'))
print("나는 {0}가 {1}개 있다.".format("사과", 2))

나는 사과가 2개 있다.
나는 사과가 두개 있다.
나는 사과가 2개 있다.


* 문자열 `format()` 메소드의 인덱스와 변수 이름을 이용한 포맷팅
* 변수 이름 s와 n을 지정하여 "사과"와 숫자 2를 매핑

In [76]:
print("나는 사과가 {n}개 있다.".format(n = 2))
print("나는 사과가 {0}개 있다.".format('두'))
print("나는 {s}가 {n}개 있다.".format(s= "사과", n = 2))

나는 사과가 2개 있다.
나는 사과가 두개 있다.
나는 사과가 2개 있다.


* 문자열 `format()` 함수의 정렬, 공백, 소수점


In [81]:
print("{0:8}".format("Hello")) # 길이 8로 하고, Hello 넣음
print("{0:<8}".format("Hello"))  # 기본 왼쪽 정렬
print("{0:>8}".format("Hello"))  # 오른쪽 정렬
print("{0:^8}".format("Hello"))  # 가운데 정렬
print("{0:~^8}".format("Hello")) # 공백을 tilt로 채움
print("{0:-^8}".format("Hello"))  # 공백을 -로 채움
print("{0:0.2f}".format(3.14159213))
print("{0:8.4f}".format(3.14159213)) # 8칸, 소수점 4자리까지
print("{{중괄호}}".format()) # 중괄호 표현

Hello   
Hello   
   Hello
 Hello  
~Hello~~
-Hello--
3.14
  3.1416
{중괄호}


### f 문자열 포맷팅

* 문자열 포맷팅을 위해 새롭게 나온 기능으로 문자열 앞에 f를 붙여서 구분
* 변수 s와 n에 대해서 문자열에 포함된 이름으로 매핑하여 사용 가능

In [83]:
s = "사과"
n = 2
print(f'나는 사과가 {n}개 있다.')
print(f'나느 {s}가 {n}개 있다.')

나는 사과가 2개 있다.
나느 사과가 2개 있다.


* 기존의 문자열 포맷을 문자열 다음 기호 : 뒤에 표현하여 그대로 사용 가능

In [87]:
print(f'{"Hello":8}') # 길이 8로 하고, Hello 넣음
print(f'{"Hello":<8}')  # 기본 왼쪽 정렬
print(f'{"Hello":>8}')  # 오른쪽 정렬
print(f'{"Hello":^8}')  # 가운데 정렬
print(f'{"Hello":~^8}') # 공백을 tilt로 채움
print(f'{"Hello":-^8}')  # 공백을 -로 채움
print(f'{3.14159214:0.2f}')
print(f'{3.14159214:8.4f}') # 8칸, 소수점 4자리까지
print(f'{{중괄호}}') 

Hello   
Hello   
   Hello
 Hello  
~Hello~~
-Hello--
3.14
  3.1416
{중괄호}
